keps/sig-testing/2420-reducing-kubernetes-build-maintenance/README.md

```diff
@@ -0,0 +1,671 @@
+  - Kubernetes has enabled repeated runs of "unit" tests to surface flakes, which causes both `go` and Bazel to not cache test results
+  - We've disabled caching for large, poorly cachable objects like Docker images anyhow
+
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+1. Continue maintaining both build systems
+
+This has major drawbacks:
+
+- This continues to eat developer time without significant return, for the most part releases, the bulk of CI (periodic tests), and contributor development use the make build(s).
+
+- Key components of our development setup do not easily port to bazel, so it will continue to be an "also":
+  - code generators: Kubernetes uses a lot of generated code, while Bazel *can* run code generators just fine without checking in the generated sources, we need to check them in for consumption by non-bazel users (i.e. external projects), and as a result we have never leveraged this. These generators are largely incompatible with bazel "philosophy":
+    - much of the k8s codegen uses fake go build tags, which would require an external process (i.e. gazelle extension) to turn into build files
+    - additionally, some codegen relies on certain dependencies, which the gazelle extension would need to figure out
+    - many of the code generators are optimized to load the entire Go tree, then generate lots of code at once, which is somewhat incompatible with Bazel's approach to working on a package-by-package level. Generating code package-by-package is much slower (due to having to reparse the tree each time).
+  - separate "hack/tools" go module for linters etc: not only do linters not work well (because any source change busts the cache anyhow), but rules_go does not do multiple go modules well. Multiple go modules allows us to isolate development dependencies (like linters) from release binary dependencies, easing dependency management
+
+2. Improve bazel integration and drop the make based build
+
+In addition to the points made in 1.) above as to why this is not particularly viable for some of our existing development patterns:
+- We would need to improve support for CGO or eliminate CGO in Kubernetes, both of which would be somewhat expensive to develop
+  - At minimum kubelet definitely requires CGO for OS integration (e.g. selinux)
+  - CGO pkg-config directives do not work in `rules_go`, requiring brittle work-arounds
+  - Cross compiling with CGO under bazel is tricky, and despite @ixdy getting close at one point, never shipped in Kubernetes, let alone portably, or capable of shipping a full release.
+    - @ixdy suspects this would largely have to be reimplemented now
+- This is unlikely to be popular with contributors
```

Comment:
That doesn't seem like a real problem? If contributors have technical issues with the Bazel workflow, they should be outlined here explicitly. If these are non-technical (eg. dislike for the sake of dislike), then these probably shouldn't be taken into account when making this decision.

Reply:
SIG Testing serves to improve the workflow of the project's contributors.
SIG Release exists to ensure we provide high quality binary releases.

Reducing developer friction is absolutely a concern. Our developers are not happy with this system (see previous discussion).

Reply:
Sure, but then these friction issues should IMO be explicitly stated within this KEP. It should obviously show that Bazel-only is a no-go.

Without knowing what the current pain points with Bazel are (and not these with two buildsystems in parallel!), it's impossible to discuss specifics.

Reply:
I would re-frame as: this is less likely to grow the amount of potential contributor bandwidth available to improve the kubernetes/kubernetes build system.  We have a larger pool of contributors today who have demonstrated experience updating our existing make build system vs. updating bazel's components and our usage of it.

Reply:
https://github.com/kubernetes/kubernetes/issues/88553 contains links to specific technical issues we've had in the past with bazel, I don't think it's necessary to re-enumerate here

Reply:
Thank you @spiffxp, that's the intent of this admittedly poorly written bullet, the "popularity" contest is surely not relevant, but the availability of experience and support within the project is / interest in doing so. I also think https://github.com/kubernetes/kubernetes/issues/88553 has sufficient technical details and is referenced in the doc.

I will update this bullet.

Reply:
This has been replaced with a variant of @spiffxp's comment.